3.7/graphs-satellite-graphs.md

```diff
@@ -0,0 +1,134 @@
+
+Traversals on graphs:
+ Id  Depth  Vertex collections  Edge collections  Options                                  Filter / Prune Conditions
+ 3   1..1   vertices            edges             uniqueVertices: none, uniqueEdges: path                           
+
+Optimization rules applied:
+ Id   RuleName
+  1   scatter-in-cluster
+  2   remove-unnecessary-remote-scatter
+```
+
+You can see that the `TraversalNode` (number 3) is executed on a Coordinator, and only
+the `EnumerateCollectionNode` is executed on DB-Server. This will happen for each of the 
+8 shards in `collection`.
+
+Let's now have a look at the same query using satellite graphs:
+
+```
+arangosh> explain(`FOR doc in collection FOR v,e,p IN OUTBOUND "vertices/start" GRAPH "satelliteGraph" RETURN [doc,v,e,p]`)
+
+Query String (102 chars, cacheable: true):
+ FOR doc in collection FOR v,e,p IN OUTBOUND "vertices/start" GRAPH "satelliteGraph" RETURN 
+ [doc,v,e,p]
+
+Execution plan:
+ Id   NodeType                  Site  Est.   Comment
+  1   SingletonNode             DBS      1   * ROOT
+  2   EnumerateCollectionNode   DBS      0     - FOR doc IN collection   /* full collection scan, 8 shard(s) */
+ 10   TraversalNode             DBS      0       - FOR v  /* vertex */, e  /* edge */, p  /* paths */ IN 1..1  /* min..maxPathDepth */ OUTBOUND 'vertices/start' /* startnode */  edges
+ 13   RemoteNode                COOR     0         - REMOTE
+ 14   GatherNode                COOR     0         - GATHER 
+  4   CalculationNode           COOR     0         - LET #6 = [ doc, v, e, p ]   /* simple expression */   /* collections used: doc : collection */
+  5   ReturnNode                COOR     0         - RETURN #6
+
+Indexes used:
+ By   Name   Type   Collection   Unique   Sparse   Selectivity   Fields        Ranges
+ 10   edge   edge   edges        false    false       100.00 %   [ `_from` ]   base OUTBOUND
+
+Traversals on graphs:
+ Id  Depth  Vertex collections  Edge collections  Options                                  Filter / Prune Conditions
+ 10  1..1                       edges             uniqueVertices: none, uniqueEdges: path                           
+
+Optimization rules applied:
+ Id   RuleName
+  1   scatter-in-cluster
+  2   remove-unnecessary-remote-scatter
+  3   scatter-satellite-graphs
+  4   remove-satellite-joins
+```
+
+Note that now the `TraversalNode` (with id 10) is executed on each DBServer, leading to a great reduction in
+required network communication, and hence potential gains in query performance.
+
+Caveats
+-------
+
+The cluster will automatically distribute all satellite graphs to all DBServers, and apply
+synchronous replication. 
+This means that writes are executed on the leader only, and then have to be replicated
+to all followers, leading to degradation in write performance.
+
+There is also a chance that a write has not been replicated in time before a query is executed
+and a query involving satellite graphs returns out of date results.
```

Comment:
If writes are done while a query is being instantiated, there is chance that a write has not been fully replicated. This can lead to inconsistent results.
```suggestion
If writes are done to a satellite graph while a query is being instantiated, there is chance that a write has not been fully replicated. This can lead to inconsistent results for queries involving that graph.
```

Reply:
Whats the difference between "instantiation" and "execution"?

I neither like "inconsistent" (because it doesn't say what it's inconsistent *with*), nor "out of date".

Reply:
> Whats the difference between "instantiation" and "execution"?

Technically, after the planning stage, the query snippets get sent to all DBServers. There, they are instantiated, meaning an object is created in the query registry. Most importantly, the local transactions are opened, and thus RocksDB Snapshots are taken on each DBServer.

Only if writes happen while the transactions are started, it can happen that one DBServer gets a write before the transaction is opened, and another does not.

> I neither like "inconsistent" (because it doesn't say what it's inconsistent with), nor "out of date".

What I meant with inconsistent, and I agree that we should clarify that, is that the different replicas of the same satellite collection(s) have a different state (snapshot) at the very beginning of the query. Thus the same traversal will return different results on different DBServers. That can lead to all kind of breakage, depending on the query. We could even say this is undefined behaviour, as of course all optimizer rules assume this does not happen.